### ISSUE TITLE
Check Servers 

### ISSUE THREAD

**Issue by ranatrk** [threefoldtech/jumpscaleX_core#30]
[development]
in jumpscaleX_core

- [x] corex
- [ ] dns
- [x] gedis
- [ ] gedis_websocket
- [x] gevent_rack
- [ ] graphql
- [ ] gundb
- [x] myjobs
- [x] openresty
- [ ] sockexec
- [x] sonic
- [x] threebot
- [ ] tmux
- [x] zdb

in jumpscaleX_libs_extra

- [x] email_forwarder
- [x] errbot
- [x] etcd
- [ ] ~~flaskserver~~
- [ ] ~~grid_capacity~~
- [ ] odoo
- [ ] radicale
- [ ] ~~raft~~
- [ ] sanic

**Comment by ranatrk** [comment]
**Corex** server done:
_install_, _check_, _start_, _stop_ functions working

fixes:
https://github.com/threefoldtech/jumpscaleX_builders/commit/cd8c0def24fd6be2e095d028db96358bd59effa4
https://github.com/threefoldtech/jumpscaleX_core/commit/d27a6272659c6590f189e4056facd7bbe2fbe3d8

**Comment by ranatrk** [comment]
**openresty** and **sonic** working

**Comment by ranatrk** [comment]
Tmux not working:
![Screenshot from 2019-09-03 14-02-10](https://user-images.githubusercontent.com/17128393/64171675-8033da00-ce53-11e9-84f4-7a90b2118d48.png)

**Comment by ranatrk** [comment]
**zdb working** (start,stop)

start kosmos using `kosmos -p` for server to start in tmux
`j.servers.zdb.get().start()`
`j.servers.zdb.get().stop()`

**Comment by ranatrk** [comment]
**Tmux** working 
to start: `j.servers.tmux`

**Comment by ranatrk** [comment]
mail_forwarder issue:
`j.servers.mail_forwarder.get()`

![Screenshot from 2019-09-03 14-29-34](https://user-images.githubusercontent.com/17128393/64173328-511f6780-ce57-11e9-95c3-9241f5e57eb7.png)

**Comment by ranatrk** [comment]
errbot issue:
need to add `install()` to install errbot module
![Screenshot from 2019-09-03 14-35-36](https://user-images.githubusercontent.com/17128393/64173722-30a3dd00-ce58-11e9-8be7-d86d3dbef59e.png)

**Comment by ranatrk** [comment]
**etcd** working(install, start, stop)

`j.servers.etcd.install()`
`j.servers.etcd.start()`
`j.servers.etcd.stop()`

fixes:
https://github.com/threefoldtech/jumpscaleX_builders/commit/63f4f0eca9d847b9ad4d20b45f5bb0f147bfdd41
https://github.com/threefoldtech/jumpscaleX_libs_extra/commit/4c7d31565b6369a2d21a5a479eb1b1d7f1596884

**Comment by ranatrk** [comment]
usage of **flask server** is deprecated, needs refactoring to be used.

**Comment by ranatrk** [comment]
>     * [ ]  ~~grid_capacity~~
**Capacity** server needs refactoring to be used

**Comment by ranatrk** [comment]
currently working on odoo and radicale servers, not working yet

**Comment by ranatrk** [comment]
>     * [ ] ~~ raft~~
**Raft** needs refactoring to be used

**Comment by ranatrk** [comment]
**Threebot** working (install, start, stop)
`j.servers.threebot.install()`

`j.servers.threebot.get().start()` or if in background 
`j.servers.threebot.get().start(background=True)`

`j.servers.threebot.get().stop()`

**Comment by ranatrk** [comment]
>     * graphql

**Graphql** start and stop not implemented!!

**Comment by ranatrk** [comment]
**Rack** and **Gedis** working:
```

**Comment by ranatrk** [comment]
**Gundb** working 
```
gundb_server = j.servers.gundb.gevent_server_get()
gundb_server.start()
```

**Comment by ranatrk** [comment]
**myjobs** working (start,stop)
` j.servers.myjobs.get().start()`
` j.servers.myjobs.get().stop(hard=True) `

**Comment by ranatrk** [comment]
working on odoo server( blocked on fixes in odoo/postgres builders)

**Comment by ranatrk** [comment]
**Graphql** server working
https://github.com/threefoldtech/jumpscaleX_core/commit/664469b550862281ba21fe6ffb5cda4781d95f32

follow docs: https://github.com/threefoldtech/jumpscaleX_core/blob/664469b550862281ba21fe6ffb5cda4781d95f32/JumpscaleCore/servers/graphql/README.md

**Comment by ranatrk** [comment]
**radicale** working 
`j.servers.radicale.default.start()`

Fix:
https://github.com/threefoldtech/jumpscaleX_libs_extra/commit/edf4a5e6b35abe64f4743fb7b4d4289400760f36

**Comment by ranatrk** [comment]
**odoo** working 
`j.servers.odoo.get().start()`

Fixes:
https://github.com/threefoldtech/jumpscaleX_libs_extra/commit/e766a0ea1b63999b327b153ea2b0f30b834c399e
https://github.com/threefoldtech/jumpscaleX_core/commit/8ed607610efaa757cc8f976abfd923edf448a840
https://github.com/threefoldtech/jumpscaleX_builders/commit/42d3dbcfeb356ee9a0415343b1b0346358fc155c

**Comment by ranatrk** [comment]
**errbot** import issue fixed

Fix:
https://github.com/threefoldtech/jumpscaleX_libs_extra/commit/e5eb42f024739f555db9ce5cda0fd248a19ee96a

**Comment by ranatrk** [comment]
**sanic** needs refactoring

**Comment by ranatrk** [comment]
Done:
- corex
- gedis
- gevent_rack
- graphql
- gundb
- myjobs
- openresty
- sonic
- threebot
- tmux
- zdb
- email_forwarder
- errbot
- etcd
- odoo
- radicale

Needs refactoring:
- flaskserver
- grid_capacity
- raft
- sanic

**Comment by Dinaamagdy** [comment]
check option in some of servers not clear need to specify :
- should check  be in all servers or not
-  what check should exactly do to make test depend on it

**Comment by Dinaamagdy** [comment]
gedis_server : 
can't do start its stuck on start and don't do anything

**Comment by Dinaamagdy** [comment]
gedis_websocket: doesn't solved yet

**Comment by Dinaamagdy** [comment]
dns :still not exist
```
JSX> j.servers.dns.get()                                                                             
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/sandbox/lib/jumpscale/Jumpscale/core/BASECLASSES/JSFactory.py", line 149, in get
    raise j.exceptions.Value("cannot get child with name:%s" % name)
Jumpscale.core.InstallTools.JSExceptions.__init__.<locals>.Value1: ?[1;31mEXCEPTION: 
    cannot get child with name:main?[0;0m

?[1;31mEXCEPTION: 
    cannot get child with name:main?[0;0m

```

**Comment by Dinaamagdy** [comment]
## graphql:
only have _server_test_start  which give the following error why it doesn't have same framework like  other server install and get method then (start- stop-delete )
```
JSX> j.servers.graphql._server_test_start()                                     
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/sandbox/lib/jumpscale/Jumpscale/servers/graphql/GraphQLFactory.py", line 56, in _server_test_start
    rack = j.servers.rack.get()
  File "/sandbox/lib/jumpscale/jumpscale_generated.py", line 1215, in rack
    from Jumpscale.servers.gevent_rack.ServerRackFactory import ServerRackFactory
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 152, in __exit__
  File "<frozen importlib._bootstrap>", line 107, in release
RuntimeError: cannot release un-acquired lock

````
## gundb
same as graphql
```
JSX> j.servers.gundb._server_test_start()                                                                                                                                                                   
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/sandbox/lib/jumpscale/Jumpscale/servers/gundb/GundbFactory.py", line 46, in _server_test_start
    rack = j.servers.rack.get()
  File "/sandbox/lib/jumpscale/jumpscale_generated.py", line 1215, in rack
    from Jumpscale.servers.gevent_rack.ServerRackFactory import ServerRackFactory
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 152, in __exit__
  File "<frozen importlib._bootstrap>", line 107, in release
RuntimeError: cannot release un-acquired lock

cannot release un-acquired lock
```
## my jobs
```
 JSX> j.servers.myjobs.get()                                                                                                                                                                                 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/sandbox/lib/jumpscale/Jumpscale/core/BASECLASSES/JSFactory.py", line 149, in get
    raise j.exceptions.Value("cannot get child with name:%s" % name)
Jumpscale.core.InstallTools.JSExceptions.__init__.<locals>.Value1: ?[1;31mEXCEPTION: 
    cannot get child with name:main?[0;0m

?[1;31mEXCEPTION: 
    cannot get child with name:main?[0;0m
```

**Comment by ranatrk** [comment]
> ## Mail forward
> 
>     * there is no install method
> 
There is no installation required

>     * Start
> 
> 
> ```
> _ Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/sandbox/lib/jumpscale/JumpscaleLibsExtra/servers/email_forwarder/JSMailForwarder.py", line 56, in start
>     self._forwarder.run()
>   File "/sandbox/lib/jumpscale/JumpscaleLibsExtra/servers/email_forwarder/forwarder.py", line 133, in run
>     self.serve_forever()
>   File "/usr/local/lib/python3.6/dist-packages/gevent/baseserver.py", line 367, in serve_forever
>     self.start()
>   File "/usr/local/lib/python3.6/dist-packages/gevent/baseserver.py", line 305, in start
>     self.init_socket()
>   File "/usr/local/lib/python3.6/dist-packages/gevent/server.py", line 146, in init_socket
>     self.socket = self.get_listener(self.address, self.backlog, self.family)
>   File "/usr/local/lib/python3.6/dist-packages/gevent/server.py", line 157, in get_listener
>     return _tcp_listener(address, backlog=backlog, reuse_addr=cls.reuse_addr, family=family)
>   File "/usr/local/lib/python3.6/dist-packages/gevent/server.py", line 256, in _tcp_listener
>     sock.bind(address)
> OverflowError: getsockaddrarg: port must be 0-65535.
> 
> getsockaddrarg: port must be 0-65535.
> JSX>  
> ```
To start the server instance should have the params needed, this is the example in the test used:
```
mf = j.servers.mail_forwarder.get(
    name="test",
    listening_host="localhost",
    listening_port=8823,
    relay_host="smtp.gmail.com",
    relay_port=587,
    relay_user="user_test",
    relay_password="password",
    relay_ssl=True,
)
mf.add_forward_config("codescalers.com", ["incubaide.com", "threefoldtech.com"])
mf.start()
```

**Comment by ranatrk** [comment]
> gedis_websocket: haven't solved yet

start already working
added stop 

 ```
server = j.servers.gedis_websocket.default 
server.start()
server.stop()
```

**Comment by ranatrk** [comment]
> gedis

gedis start and stop adjusted to start/stop in tmux

**Comment by ranatrk** [comment]
> ## graphql:
> 
> only have _server_test_start which give the following error why it doesn't have same framework like other server install and get method then (start- stop-delete )
> 
> ```
> JSX> j.servers.graphql._server_test_start()                                     
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/sandbox/lib/jumpscale/Jumpscale/servers/graphql/GraphQLFactory.py", line 56, in _server_test_start
>     rack = j.servers.rack.get()
>   File "/sandbox/lib/jumpscale/jumpscale_generated.py", line 1215, in rack
>     from Jumpscale.servers.gevent_rack.ServerRackFactory import ServerRackFactory
>   File "<frozen importlib._bootstrap>", line 971, in _find_and_load
>   File "<frozen importlib._bootstrap>", line 152, in __exit__
>   File "<frozen importlib._bootstrap>", line 107, in release
> RuntimeError: cannot release un-acquired lock
> ```

@Dinaamagdy run kosmos using  `kosmos -p` then try again

**Comment by Dinaamagdy** [comment]
odoo server : https://github.com/threefoldtech/jumpscaleX_core/issues/92

**Comment by Dinaamagdy** [comment]
tmux server : https://github.com/threefoldtech/jumpscaleX_core/issues/93

**Comment by Dinaamagdy** [comment]
## zdb 
- zdb should have install and now it doesn't have

**Comment by Dinaamagdy** [comment]
I create separate issue for servers which need refactor 
https://github.com/threefoldtech/jumpscaleX_core/issues/94

**Comment by ranatrk** [comment]
> ## zdb
> 
>     * zdb should have install and now it doesn't have
> 
> 
> ```
> JSX> j.servers.zdb.install()                                                                                                                                                                                
> --TRACEBACK------------------
> <stdin> in <module>
>     1    
> /sandbox/lib/jumpscale/Jumpscale/core/BASECLASSES/Attr.py in __getattr__
>     57    % (name, self._key)
> -----------------------------
> Wed 18 11:38:00 /BASECLASSES/Attr.py -  57 - __getattr__                        : EXCEPTION: 
>     try to get attribute: 'install', instance did not exist, was also not a method or property, was on 'zdbservers'
> ```

changed build signature to install

**Comment by Dinaamagdy** [comment]
## errbot
- need to separate install method 
- need to add stop method
- start return the following+ error :
```
 uccessfully installed WebOb-1.8.5 Werkzeug-0.15.6 ansi-0.1.3 beautifulsoup4-4.8.0 certifi-2019.9.11 click-7.0 daemonize-2.5.0 dataclasses-0.6 dulwich-0.19.13 errbot-6.1.1 idna-2.8 markdown-2.6.11 setuptools-41.2.0 soupsieve-1.9.3 waitress-1.3.1 webtest-2.0.33
16:23:24 INFO     errbot.bootstrap          Found Storage plugin: Shelf.
16:23:24 INFO     errbot.bootstrap          Found Backend plugin: Telegram
16:23:24 ERROR    errbot.backends.telegram_ Could not start the Telegram back-end
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/errbot/backends/telegram_messenger.py", line 15, in <module>
    import telegram
ModuleNotFoundError: No module named 'telegram'
16:23:24 CRITICAL errbot.backends.telegram_ You need to install the telegram support in order to use the Telegram backend.
You should be able to install this package using:
pip install errbot[telegram]
```

**Comment by ranatrk** [comment]
> ## errbot
> 
>     * need to separate install method
> 
>     * need to add stop method
> 
>     * start return the following+ error :
> 
> 
> ```
>  uccessfully installed WebOb-1.8.5 Werkzeug-0.15.6 ansi-0.1.3 beautifulsoup4-4.8.0 certifi-2019.9.11 click-7.0 daemonize-2.5.0 dataclasses-0.6 dulwich-0.19.13 errbot-6.1.1 idna-2.8 markdown-2.6.11 setuptools-41.2.0 soupsieve-1.9.3 waitress-1.3.1 webtest-2.0.33
> 16:23:24 INFO     errbot.bootstrap          Found Storage plugin: Shelf.
> 16:23:24 INFO     errbot.bootstrap          Found Backend plugin: Telegram
> 16:23:24 ERROR    errbot.backends.telegram_ Could not start the Telegram back-end
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.6/dist-packages/errbot/backends/telegram_messenger.py", line 15, in <module>
>     import telegram
> ModuleNotFoundError: No module named 'telegram'
> 16:23:24 CRITICAL errbot.backends.telegram_ You need to install the telegram support in order to use the Telegram backend.
> You should be able to install this package using:
> pip install errbot[telegram]
> ```

refactored errbot server, now install is seperate

**Comment by Dinaamagdy** [comment]
https://github.com/threefoldtech/jumpscaleX_core/issues/101

**Comment by Dinaamagdy** [comment]
https://github.com/threefoldtech/jumpscaleX_core/issues/104

**Comment by Dinaamagdy** [comment]
https://github.com/threefoldtech/jumpscaleX_core/issues/109

**Comment by Dinaamagdy** [comment]
sockexec : 
get he follwing error 
```Traceback (most recent call last): 
  File "/sandbox/lib/jumpscale/Jumpscale/core/KosmosShell.py", line 456, in custom_get_completions 
    completions = list(get_completions(self, document, complete_event)) 
  File "/sandbox/lib/jumpscale/Jumpscale/core/KosmosShell.py", line 115, in get_completions 
    obj = eval_code(parent, self.get_locals(), self.get_globals()) 
  File "/sandbox/lib/jumpscale/Jumpscale/core/KosmosShell.py", line 46, in eval_code 
    return eval(code, globals_, locals_) 
  File "<kosmos>", line 1, in <module> 
  File "/sandbox/lib/jumpscale/jumpscale_generated.py", line 1197, in sockexec 
    from Jumpscale.servers.sockexec.SockExec import SockExec 
  File "/sandbox/lib/jumpscale/Jumpscale/servers/sockexec/SockExec.py", line 8, in <module> 
    class SockExec(j.application.JSBaseDataObjClass): 
AttributeError: 'Application' object has no attribute 'JSBaseDataObjClass' 
 
```

**Comment by Dinaamagdy** [comment]
gedis_websocket:
start : not works sucessfully as I cant get its process

**Comment by Dinaamagdy** [comment]
https://github.com/threefoldtech/jumpscaleX_core/issues/126

**Comment by ranatrk** [comment]
> sockexec :
> get he follwing error
> 
> ```
>   File "/sandbox/lib/jumpscale/Jumpscale/core/KosmosShell.py", line 456, in custom_get_completions 
>     completions = list(get_completions(self, document, complete_event)) 
>   File "/sandbox/lib/jumpscale/Jumpscale/core/KosmosShell.py", line 115, in get_completions 
>     obj = eval_code(parent, self.get_locals(), self.get_globals()) 
>   File "/sandbox/lib/jumpscale/Jumpscale/core/KosmosShell.py", line 46, in eval_code 
>     return eval(code, globals_, locals_) 
>   File "<kosmos>", line 1, in <module> 
>   File "/sandbox/lib/jumpscale/jumpscale_generated.py", line 1197, in sockexec 
>     from Jumpscale.servers.sockexec.SockExec import SockExec 
>   File "/sandbox/lib/jumpscale/Jumpscale/servers/sockexec/SockExec.py", line 8, in <module> 
>     class SockExec(j.application.JSBaseDataObjClass): 
> AttributeError: 'Application' object has no attribute 'JSBaseDataObjClass' 
>  
> ```

Fix : https://github.com/threefoldtech/jumpscaleX_core/commit/1fd82a76ba53f3c212dbed945d7ca6c8ad6f12ab
